share/eds/store.go

```diff
@@ -0,0 +1,211 @@
+	key := root.String()
+	f, err := os.OpenFile(s.basepath+blocksPath+key, os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+
+	err = WriteEDS(ctx, square, f)
+	if err != nil {
+		return err
+	}
+
+	ch := make(chan dagstore.ShardResult, 1)
+	err = s.dgstr.RegisterShard(ctx, shard.KeyFromString(key), &mount.FSMount{
+		FS:   os.DirFS(s.basepath + blocksPath),
+		Path: key,
+	}, ch, dagstore.RegisterOpts{})
+
+	if err != nil {
+		return err
+	}
+
+	result := <-ch
+	if result.Error != nil {
+		return result.Error
+	}
+
+	return nil
+}
+
+// GetCAR takes a DataRoot and returns a buffered reader to the respective EDS serialized as a CARv1 file.
+//
+// The Reader strictly reads the first quadrant(1/4) of EDS, omitting all the NMT Merkle proofs.
+// Integrity of the store data is not verified.
+//
+// Caller must Close returned reader after reading.
+func (s *EDSStore) GetCAR(ctx context.Context, root share.Root) (io.ReadCloser, error) {
+	key := root.String()
+
+	ch := make(chan dagstore.ShardResult, 1)
+	err := s.dgstr.AcquireShard(ctx, shard.KeyFromString(key), ch, dagstore.AcquireOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	result := <-ch
+	if result.Error != nil {
+		return nil, result.Error
+	}
+	return result.Accessor, nil
+}
+
+// Blockstore returns an IPFS Blockstore providing access to individual shares/nodes of all EDS
+// registered on the Store. NOTE: The Blockstore does not store whole Celestia Blocks but IPFS blocks.
+// We represent `shares` and NMT Merkle proofs as IPFS blocks and IPLD nodes so Bitswap can access those.
+func (s *EDSStore) Blockstore() blockstore.Blockstore {
+	return s.bs
+}
+
+// Remove removes EDS from Store by the given share.Root and cleans up all the indexing.
+func (s *EDSStore) Remove(ctx context.Context, root share.Root) error {
+	key := root.String()
+	ch := make(chan dagstore.ShardResult, 1)
+	err := s.dgstr.DestroyShard(ctx, shard.KeyFromString(key), ch, dagstore.DestroyOpts{})
+	if err != nil {
+		return err
+	}
+
+	result := <-ch
+	if result.Error != nil {
+		return result.Error
+	}
+
+	dropped, err := s.carIdx.DropFullIndex(shard.KeyFromString(key))
+	if !dropped {
+		log.Warnf("failed to drop index for %s", key)
+	}
+	return err
+}
+
+// Get reads EDS out of Store by given DataRoot.
+//
+// It reads only one quadrant(1/4) of the EDS and verifies the integrity of the stored data by recomputing it.
+func (s *EDSStore) Get(ctx context.Context, root share.Root) (*rsmt2d.ExtendedDataSquare, error) {
+	key := root.String()
+	f, err := os.OpenFile(s.basepath+blocksPath+key, os.O_RDONLY, 0600)
+	if err != nil {
+		return nil, err
+	}
+	return ReadEDS(ctx, f, root)
+}
+
+// Has checks if EDS exists by the given share.Root.
+func (s *EDSStore) Has(ctx context.Context, root share.Root) (bool, error) {
+	key := root.String()
+	info, err := s.dgstr.GetShardInfo(shard.KeyFromString(key))
+	if err == dagstore.ErrShardUnknown {
+		return false, err
+	}
+
+	return true, info.Error
```

Comment:
i think it is a bit more nuanced than this because shards have different states than just exists and not exists. for example the dagstore can know that the dagstore exists but is not retrievable (but i think it doesn't apply to FS mount). will look more into this tomorrow